toolchain/check/deduce.cpp

```diff
@@ -208,65 +266,99 @@ auto DeductionContext::Deduce() -> bool {
-                result_arg_ids_[index.index] != arg_const_inst_id) {
-              if (diagnose_) {
-                // TODO: Include the two different deduced values.
-                CARBON_DIAGNOSTIC(
-                    DeductionInconsistent, Error,
-                    "inconsistent deductions for value of generic "
-                    "parameter `{0}`",
-                    SemIR::NameId);
-                auto diag = context().emitter().Build(
-                    loc_id_, DeductionInconsistent, entity_name.name_id);
-                NoteGenericHere(context(), generic_id_, diag);
-                diag.Emit();
-              }
-              return false;
+        if (!index.is_valid() || index < first_deduced_index_) {
+          break;
+        }
+
+        CARBON_CHECK(static_cast<size_t>(index.index) < result_arg_ids_.size(),
+                     "Deduced value for unexpected index {0}; expected to "
+                     "deduce {1} arguments.",
+                     index, result_arg_ids_.size());
+        auto arg_const_inst_id =
+            context().constant_values().GetConstantInstId(arg_id);
+        if (arg_const_inst_id.is_valid()) {
+          if (result_arg_ids_[index.index].is_valid() &&
+              result_arg_ids_[index.index] != arg_const_inst_id) {
+            if (diagnose_) {
+              // TODO: Include the two different deduced values.
+              CARBON_DIAGNOSTIC(DeductionInconsistent, Error,
+                                "inconsistent deductions for value of generic "
+                                "parameter `{0}`",
+                                SemIR::NameId);
+              auto diag = context().emitter().Build(
+                  loc_id_, DeductionInconsistent, entity_name.name_id);
+              NoteGenericHere(context(), generic_id_, diag);
+              diag.Emit();
             }
-            result_arg_ids_[index.index] = arg_const_inst_id;
+            return false;
           }
+          result_arg_ids_[index.index] = arg_const_inst_id;
         }
-        break;
+        continue;
       }
 
+      // Various kinds of parameter should match an argument of the same form,
+      // if the operands all match.
+      case SemIR::ArrayType::Kind:
+      case SemIR::ClassType::Kind:
+      case SemIR::ConstType::Kind:
+      case SemIR::FloatType::Kind:
+      case SemIR::InterfaceType::Kind:
+      case SemIR::IntType::Kind:
+      case SemIR::PointerType::Kind:
+      case SemIR::TupleType::Kind:
+      case SemIR::TupleValue::Kind: {
+        auto arg_inst = context().insts().Get(arg_id);
+        if (arg_inst.kind() != param_inst.kind()) {
+          break;
+        }
+        auto [kind0, kind1] = param_inst.ArgKinds();
+        worklist_.AddInstArg(kind0, param_inst.arg0(), arg_inst.arg0(),
+                             needs_substitution);
+        worklist_.AddInstArg(kind1, param_inst.arg1(), arg_inst.arg1(),
+                             needs_substitution);
```

Comment:
Is it worth at all conditioning this based on the # of args to the inst? Or if the arg is invalid?

Reply:
We could have different case blocks for different numbers of arguments, but I think it's *probably* not worth the complexity and tighter coupling. The switch in `AddInstArg` immediately catches the case of no corresponding argument, so we should deal with that case quickly.

So far, we don't have any cases where the arg can be present but invalid. I guess we can handle that if/when it starts coming up.